src/extrema_set.rs

```diff
@@ -0,0 +1,42 @@
+/// Implementation guts for `min_set`, `min_set_by`, and `min_set_by_key`.
+pub fn min_set_impl<I, K, F, L>(mut it: I,
+                                mut key_for: F,
+                                mut lt: L) -> Option<Vec<I::Item>>
+    where I: Iterator,
+          F: FnMut(&I::Item) -> K,
+          L: FnMut(&I::Item, &I::Item, &K, &K) -> bool,
+{
+    let (mut result, mut current_key) = match it.next() {
+        None => return None,
+        Some(element) => {
+            let key = key_for(&element);
+            (vec![element], key)
+        }
+    };
+
+    for element in it {
+        let key = key_for(&element);
+        if lt(&element, &result[0], &key, &current_key) {
+            result.clear();
```

Comment:
Referring to https://github.com/rust-itertools/itertools/pull/323#discussion_r310403165:

> Since the code by neccesity (unless double iteration is used) _provisionally_ accumulates an unbounded number of result items, and discards these when a new provisionally best results is found, something more than `FromIterator + Extend` would be needed AFAIK.
> 
> My guess would be that anything that also has the possibility to clear the results is no better than the interal `Vec`, and would also be less clear. I'd be glad to learn I'm wrong though!

I guess you are right that a `Vec` is the best choice for many (if not most) cases. However, if someone deliberately wants e.g. a `HashSet`, I think we ideally want to avoid the conversion of a `Vec` to a `HashSet`. I.e. I suggest that `max_set` should be usable somewhat like to `collect`:

```
let v = vec![0,2,0,1,2,1,0,2,0,1,1];
let i = vec.iter().cloned();
let as_vec      = i.clone().collect::<Vec<_>>();
let as_hashset  = i.clone().collect::<HashSet<_>>();
let max_vec     = i.clone().max_set::<Vec<_>>();
let max_hashset = i.clone().max_set::<HashSet<_>>(); // should imho work with HashSet internally
````

We could achieve that by replacing (https://github.com/rust-itertools/itertools/pull/323/commits/feea62176e8aa70c53debc5e91751a0fe7635a44#diff-1941d90b206003901f556eb26de20abcR20-R21)

```
result.clear();
result.push(element);
```

by

```
result = Some(element).into_iter().collect();
```

Similarily, we could replace (https://github.com/rust-itertools/itertools/pull/323/commits/feea62176e8aa70c53debc5e91751a0fe7635a44#diff-1941d90b206003901f556eb26de20abcR13)

```
(vec![element], key)
```

by

```
(Some(element).into_iter().collect(), key)
```

All of these would only require `FromIterator` (as far as I can see). Unfortunately, I do not know the performance implications of those changes, but I would hope that the compiler is smart enough to optimize it decently.

Reply:
By the way: Is there a simple way to quote parts of a patch? (I just copy-pasted everything manually.)